--- /Users/baylor/Research/MITgcm.checking/model/src/external_forcing.F	Tue Aug 23 13:44:16 2005
+++ external_forcing.F	Thu Sep 22 11:34:39 2005
@@ -285,6 +285,18 @@
        ENDDO
       ENDIF
 
+C     Restore Temp at depth where invtime_restore isn't zero
+C     Restoring of ksurface treated in external_forcing_surf
+      if (kLev.ne.ksurface) then 
+        DO j=jMin,jMax
+         DO i=iMin,iMax
+          gT(i,j,kLev,bi,bj)=gT(i,j,kLev,bi,bj)
+     &       +(theta_restore(i,j,kLev,bi,bj)-theta(i,j,kLev,bi,bj))*
+     &       invtime_restore(i,j,kLev,bi,bj)*maskC(i,j,kLev, bi,bj)
+          ENDDO
+         ENDDO
+      endif	
+
 #ifdef SHORTWAVE_HEATING
 C Penetrating SW radiation
 c     IF ( usePenetratingSW ) THEN
@@ -379,6 +391,18 @@
       ENDIF
 
 C--   Forcing term
+C     Restore salinity at depth where invtime_restore isn't zero
+C     Restoring of ksurface treated in external_forcing_surf
+      if (kLev.ne.ksurface) then 
+        DO j=jMin,jMax
+         DO i=iMin,iMax
+          gS(i,j,kLev,bi,bj)=gS(i,j,kLev,bi,bj)
+     &       +(spice_restore(i,j,kLev,bi,bj)-salt(i,j,kLev,bi,bj))*
+     &       invtime_restore(i,j,kLev,bi,bj)*maskC(i,j,kLev,bi,bj)
+          ENDDO
+         ENDDO
+      endif	
+
 #ifdef ALLOW_AIM
       IF ( useAIM ) CALL AIM_TENDENCY_APPLY_S(
      &                      iMin,iMax, jMin,jMax, bi,bj, kLev,
