--- ../../../pkg/kpp/KPP_OPTIONS.h	2005-09-11 12:52:08.000000000 -0400
+++ KPP_OPTIONS.h	2005-09-22 12:06:14.000000000 -0400
@@ -33,7 +33,7 @@
 C o When set, smooth zonal shear meridionally and
 C   meridional shear zonally with 121 filters
 #define KPP_SMOOTH_SHSQ
-#undef KPP_SMOOTH_DVSQ
+#define KPP_SMOOTH_DVSQ
 
 C o When set, smooth dbloc KPP variable horizontally
 #define KPP_SMOOTH_DBLOC
@@ -45,10 +45,10 @@
 #endif
 
 C o When set, smooth vertical viscosity horizontally
-#undef KPP_SMOOTH_VISC
+#define KPP_SMOOTH_VISC
 
 C o When set, smooth vertical diffusivity horizontally
-#undef KPP_SMOOTH_DIFF
+#define KPP_SMOOTH_DIFF
 
 C o Get rid of vertical resolution dependence of dVsq term by
 C   estimating a surface velocity that is independent of first
