--- ../../../model/inc/PARAMS.h	2005-12-13 10:05:23.000000000 -0500
+++ PARAMS.h	2005-10-14 10:05:37.000000000 -0400
@@ -1,4 +1,4 @@
-C $Header: /u/gcmpack/MITgcm/model/inc/PARAMS.h,v 1.172 2005/12/08 00:15:05 stephd Exp $
+C $Header: /u/gcmpack/MITgcm/model/inc/PARAMS.h,v 1.169 2005/10/11 21:27:20 jmc Exp $
 C $Name:  $
 C
 
@@ -71,6 +71,9 @@
 C     surfQnetFile    :: File containing surface net heat flux
 C     surfQswFile     :: File containing surface shortwave radiation
 C     dQdTfile        :: File containing thermal relaxation coefficient
+C     invtimeRestoreFile :: File containing inverse resotoring time
+C     thetaRestoreFile   :: File containing temperature restoring values
+C     spiceRestoreFile   :: File containing spice restoring values
 C     EmPmRfile       :: File containing surface fresh water flux
 C     saltFluxFile    :: File containing surface salt flux
 C     pLoadFile       :: File containing pressure loading
@@ -95,6 +98,8 @@
      &                eddyTauxFile, eddyTauyFile,
      &                eosType, pickupSuff,
      &                mdsioLocalDir, 
+     &                invtimeRestoreFile,thetaRestoreFile,
+     &                spiceRestoreFile,
      &                the_run_name
       CHARACTER*(5) checkPtSuff(maxNoChkptLev)
       CHARACTER*(MAX_LEN_FNAM) delXFile
@@ -123,6 +128,9 @@
       CHARACTER*(MAX_LEN_FNAM) lambdaThetaFile
       CHARACTER*(MAX_LEN_FNAM) lambdaSaltFile
       CHARACTER*(MAX_LEN_FNAM) mdsioLocalDir
+      CHARACTER*(MAX_LEN_FNAM) invtimeRestoreFile
+      CHARACTER*(MAX_LEN_FNAM) thetaRestoreFile
+      CHARACTER*(MAX_LEN_FNAM) spiceRestoreFile
       CHARACTER*(MAX_LEN_FNAM) the_run_name
       CHARACTER*(6) eosType
       CHARACTER*(10) pickupSuff
@@ -260,10 +268,8 @@
 C     momImplVertAdv  :: Turns on implicit vertical advection for Momentum
 C     multiDimAdvection :: Flag that enable multi-dimension advection
 C     useMultiDimAdvec  :: True if multi-dim advection is used at least once
-C     forcing_In_AB     :: if False, put forcing (Temp,Salt,Tracers) contribution
-C                          out off Adams-Bashforth time stepping.
-C     doAB_onGtGs       :: if the Adams-Bashforth time stepping is used, always
-C                          apply AB on tracer tendencies (rather than on Tracer) 
+C     forcing_In_AB :: if False, put forcing (Temp,Salt,Tracers) contribution
+C                      out off Adams-Bashforth time stepping.
 C     startFromPickupAB2 :: with AB-3 code, start from an AB-2 pickup 
 C     doThetaClimRelax :: Set true if relaxation to temperature
 C                        climatology is required.
@@ -323,8 +329,7 @@
      & useAbsVorticity, upwindShear,
      & implicitDiffusion, implicitViscosity,
      & tempImplVertAdv, saltImplVertAdv, momImplVertAdv,
-     & multiDimAdvection, useMultiDimAdvec, 
-     & forcing_In_AB, doAB_onGtGs,
+     & multiDimAdvection, useMultiDimAdvec, forcing_In_AB,
      & doThetaClimRelax, doSaltClimRelax, doTr1ClimRelax, 
      & periodicExternalForcing, 
      & fluidIsAir, fluidIsWater,
@@ -389,7 +394,6 @@
       LOGICAL multiDimAdvection
       LOGICAL useMultiDimAdvec
       LOGICAL forcing_In_AB
-      LOGICAL doAB_onGtGs
       LOGICAL doThetaClimRelax
       LOGICAL doSaltClimRelax
       LOGICAL doTr1ClimRelax
@@ -451,7 +455,6 @@
 C     rhoConstFresh :: Constant reference density for fresh water (rain)
 C     tRef      :: reference vertical profile for potential temperature
 C     sRef      :: reference vertical profile for salinity/specific humidity 
-C     phiRef    :: reference potential (pressure/rho, geopotential) profile 
 C     phiMin    :: Latitude of southern most cell face.
 C     thetaMin  :: Longitude of western most cell face (this
 C                 is an "inert" parameter but it is included
@@ -596,6 +599,10 @@
 C     bottomDragQuadratic   ( linear: 1/s, quadratic: 1/m )
 C     nh_Am2        :: scales the non-hydrostatic terms and changes internal scales
 C                      (i.e. allows convection at different Rayleigh numbers)
+C    fuconst             :: Constant Zonal Wind Stress
+C    daylength           :: Duration of Daytime insolation (sec)
+C    Qdiur               :: Magnitude of diurnal forcing   (W/m2)
+C    Qconst              :: Magnitude of constant nightime Cooling  (W/m2)
       COMMON /PARM_R/ cg2dTargetResidual, cg2dTargetResWunit, 
      & cg2dpcOffDFac, cg3dTargetResidual,
      & delR, delRc, delX, delY,
@@ -618,7 +625,7 @@
      & hFacMin, hFacMinDz, hFacInf, hFacSup,
      & gravity, recip_Gravity, gBaro, rhonil, recip_rhonil, 
      & recip_rhoConst, rhoConst, 
-     & rhoConstFresh, convertEmP2rUnit, tRef, sRef, phiRef,
+     & rhoConstFresh, convertEmP2rUnit, tRef, sRef,
      & baseTime, startTime, endTime, 
      & chkPtFreq, pchkPtFreq, dumpFreq, adjDumpFreq,
      & diagFreq, taveFreq, tave_lastIter, monitorFreq, adjMonitorFreq,
@@ -633,6 +640,7 @@
      & horiVertRatio, recip_horiVertRatio,
      & ivdc_kappa, Ro_SeaLevel,
      & sideDragFactor, bottomDragLinear, bottomDragQuadratic, nh_Am2,
+     & fuconst,daylength,Qdiur,Qconst,
      & tCylIn, tCylOut
 
       _RL cg2dTargetResidual
@@ -711,7 +719,6 @@
       _RL convertEmP2rUnit
       _RL tRef(Nr)
       _RL sRef(Nr)
-      _RL phiRef(2*Nr+1)
       _RL baseTime
       _RL startTime
       _RL endTime
@@ -752,6 +759,10 @@
       _RL bottomDragLinear
       _RL bottomDragQuadratic
       _RL nh_Am2
+      _RL fuconst
+      _RL daylength
+      _RL Qdiur
+      _RL Qconst
       _RL tCylIn
       _RL tCylOut
 
@@ -797,7 +808,6 @@
       LOGICAL useFLT
       LOGICAL usePTRACERS
       LOGICAL useGCHEM
-      LOGICAL useRBCS
       LOGICAL useSBO
       LOGICAL useSEAICE
       LOGICAL useBulkForce
@@ -812,7 +822,7 @@
       COMMON /PARM_PACKAGES/
      &        useKPP, useGMRedi, useOBCS, useAIM, useLand, useECCO, 
      &        useSHAP_FILT, useZONAL_FILT, useGrdchk, useFLT,
-     &        usePTRACERS, useGCHEM, useRBCS,
+     &        usePTRACERS, useGCHEM,
      &        useSBO, useSEAICE, useThSIce, useBulkForce, 
      &        usefizhi, usegridalt, usediagnostics, useEBM, useMNC,
      &        usePP81, useMY82, useOPPS, useGGL90, useMATRIX,
