--- ../../../model/inc/SIZE.h	2005-09-19 14:42:35.000000000 -0400
+++ SIZE.h	2005-10-20 22:58:05.000000000 -0400
@@ -1,13 +1,5 @@
-C $Header: /u/gcmpack/MITgcm/model/inc/SIZE.h,v 1.27 2005/09/19 18:42:35 baylor Exp $
+C $Header: /u/gcmpack/MITgcm/model/inc/SIZE.h,v 1.26 2001/09/21 15:13:31 cnh Exp $
 C $Name:  $
-
-  These lines are here to deliberately cause a compile-time error.
-  If you see these lines in your .F files or the compiler shows them
-  as an error then it means you have not placed your configuration
-  files in the appropriate place. 
-  You need to place you own copy of SIZE.h in the include
-  path for the model, and comment out these lines.
-
 C
 CBOP
 C    !ROUTINE: SIZE.h
@@ -51,17 +43,17 @@
       INTEGER Ny
       INTEGER Nr
       PARAMETER (
-     &           sNx =  20,
-     &           sNy =  16,
-     &           OLx =   4,
-     &           OLy =   4,
+     &           sNx = 150,
+     &           sNy =  30,
+     &           OLx =   3,
+     &           OLy =   3,
      &           nSx =   1,
      &           nSy =   1,
      &           nPx =   1,
-     &           nPy =   1,
+     &           nPy =   5,
      &           Nx  = sNx*nSx*nPx,
      &           Ny  = sNy*nSy*nPy,
-     &           Nr  =  23)
+     &           Nr  =  60)
 
 C     MAX_OLX  - Set to the maximum overlap region size of any array
 C     MAX_OLY    that will be exchanged. Controls the sizing of exch
