diff -Naur -x Entries -IHeader MITgcm-c64l/eesupp/src/exch1_ad.flow MITgcm-c64l-v2-patch-final/eesupp/src/exch1_ad.flow
--- MITgcm-c64l/eesupp/src/exch1_ad.flow	2010-10-16 19:01:04.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/eesupp/src/exch1_ad.flow	2013-08-16 00:56:22.000000000 +0200
@@ -1,5 +1,27 @@
-C $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-C $Name:  $
+C $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+C $Name:  $
+
+C------------------------------------------
+C directives to specify the adjoint names
+C of routines called in the original code
+C------------------------------------------
+c$taf SUBROUTINE exch_3d_rl      ADNAME = adexch_3d_rl
+c$taf SUBROUTINE exch_3d_rs      ADNAME = adexch_3d_rs
+c$taf SUBROUTINE exch_uv_3d_rl   ADNAME = adexch_uv_3d_rl
+c$taf SUBROUTINE exch_uv_3d_rs   ADNAME = adexch_uv_3d_rs
+c$taf SUBROUTINE exch_uv_xy_rl   ADNAME = adexch_uv_xy_rl
+c$taf SUBROUTINE exch_uv_xy_rs   ADNAME = adexch_uv_xy_rs
+c$taf SUBROUTINE exch_uv_xyz_rl  ADNAME = adexch_uv_xyz_rl
+c$taf SUBROUTINE exch_uv_xyz_rs  ADNAME = adexch_uv_xyz_rs
+c$taf SUBROUTINE exch_xy_rl      ADNAME = adexch_xy_rl
+c$taf SUBROUTINE exch_xy_rs      ADNAME = adexch_xy_rs
+c$taf SUBROUTINE exch_xyz_rl     ADNAME = adexch_xyz_rl
+c$taf SUBROUTINE exch_xyz_rs     ADNAME = adexch_xyz_rs
+c$taf SUBROUTINE exch_uv_agrid_3d_rl ADNAME = adexch_uv_agrid_3d_rl
+c$taf SUBROUTINE exch_uv_bgrid_3d_rl ADNAME = adexch_uv_agrid_3d_rl
+c$taf SUBROUTINE exch_uv_dgrid_3d_rl ADNAME = adexch_uv_dgrid_3d_rl
+c$taf SUBROUTINE exch_sm_3d_rl   ADNAME = adexch_sm_3d_rl
+c$taf SUBROUTINE exch_z_3d_rl    ADNAME = adexch_z_3d_rl
 
 C------------------------------------------
 C TAF flow directives for exch1_RX
diff -Naur -x Entries -IHeader MITgcm-c64l/model/src/port_rand.F MITgcm-c64l-v2-patch-final/model/src/port_rand.F
--- MITgcm-c64l/model/src/port_rand.F	2009-12-08 22:50:35.000000000 +0100
+++ MITgcm-c64l-v2-patch-final/model/src/port_rand.F	2013-08-16 00:56:32.000000000 +0200
@@ -1,5 +1,5 @@
-C $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-C $Name:  $
+C $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+C $Name:  $
 
 #undef _USE_INTEGERS
 
@@ -162,19 +162,16 @@
 c     seed=1618033.0d0
 
 C     first generate 2 equally distributed random numbers (-1,1)
-      DO WHILE (1 .EQ. 1)
+      xs = 0.0
+      DO WHILE ( xs.GE.1.0 .OR. xs.EQ.0.0 )
          x1=2.0*port_rand(seed)-1.0
          x2=2.0*port_rand(seed)-1.0
          xs=x1**2+x2**2
-         IF (xs .LT. 1.0 .AND. xs .NE. 0.0) THEN
-            GOTO 100
-         ENDIF
       ENDDO
- 100  CONTINUE
 
       t = SQRT(-2.0*LOG(xs)/xs)
       port_rand_norm = t*x1
-C
+
 C     also t*x2 would be a gaussian random number and could be returned
 
       RETURN
diff -Naur -x Entries -IHeader MITgcm-c64l/pkg/autodiff/autodiff_whtapeio_sync.flow MITgcm-c64l-v2-patch-final/pkg/autodiff/autodiff_whtapeio_sync.flow
--- MITgcm-c64l/pkg/autodiff/autodiff_whtapeio_sync.flow	2011-10-30 23:48:10.000000000 +0100
+++ MITgcm-c64l-v2-patch-final/pkg/autodiff/autodiff_whtapeio_sync.flow	2013-08-16 00:56:47.000000000 +0200
@@ -1,5 +1,5 @@
-C $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-C $Name:  $
+C $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+C $Name:  $
 
 C     /==========================================================\
 C     | SUBROUTINE autodiff_whtapeio_sync                        |
@@ -11,3 +11,4 @@
 cadj SUBROUTINE autodiff_whtapeio_sync REQUIRED
 cadj SUBROUTINE autodiff_whtapeio_sync INFLUENCED
 cadj SUBROUTINE autodiff_whtapeio_sync FTLNAME = autodiff_whtapeio_sync
+c$taf SUBROUTINE autodiff_whtapeio_sync ADNAME=adautodiff_whtapeio_sync
diff -Naur -x Entries -IHeader MITgcm-c64l/pkg/autodiff/common.flow MITgcm-c64l-v2-patch-final/pkg/autodiff/common.flow
--- MITgcm-c64l/pkg/autodiff/common.flow	1970-01-01 01:00:00.000000000 +0100
+++ MITgcm-c64l-v2-patch-final/pkg/autodiff/common.flow	2013-08-16 00:57:03.000000000 +0200
@@ -0,0 +1,104 @@
+c$taf COMMON dynvars_r  ADNAME = addynvars_r
+c$taf COMMON dynvars_r FTLNAME = g_dynvars_r
+
+c$taf COMMON dynvars_r2  ADNAME = addynvars_r2
+c$taf COMMON dynvars_r2 FTLNAME = g_dynvars_r2
+
+c$taf COMMON dynvars_diag  ADNAME = addynvars_diag
+c$taf COMMON dynvars_diag FTLNAME = g_dynvars_diag
+
+c$taf COMMON dynvars_cd  ADNAME = addynvars_cd
+c$taf COMMON dynvars_cd FTLNAME = g_dynvars_cd
+
+c$taf COMMON ffields_fu  ADNAME = adffields_fu
+c$taf COMMON ffields_fu FTLNAME = g_ffields_fu
+
+c$taf COMMON ffields_fv  ADNAME = adffields_fv
+c$taf COMMON ffields_fv FTLNAME = g_ffields_fv
+
+c$taf COMMON ffields_Qnet  ADNAME = adffields_Qnet
+c$taf COMMON ffields_Qnet FTLNAME = g_ffields_Qnet
+
+c$taf COMMON ffields_Qsw  ADNAME = adffields_Qsw
+c$taf COMMON ffields_Qsw FTLNAME = g_ffields_Qsw
+
+c$taf COMMON ffields_dQdT  ADNAME = adffields_dQdT
+c$taf COMMON ffields_dQdT FTLNAME = g_ffields_dQdT
+
+c$taf COMMON ffields_EmPmR  ADNAME = adffields_EmPmR
+c$taf COMMON ffields_EmPmR FTLNAME = g_ffields_EmPmR
+
+c$taf COMMON ffields_saltFlux  ADNAME = adffields_saltFlux
+c$taf COMMON ffields_saltFlux FTLNAME = g_ffields_saltFlux
+
+c$taf COMMON ffields_SST  ADNAME = adffields_SST
+c$taf COMMON ffields_SST FTLNAME = g_ffields_SST
+
+c$taf COMMON ffields_SSS  ADNAME = adffields_SSS
+c$taf COMMON ffields_SSS FTLNAME = g_ffields_SSS
+
+c$taf COMMON ffields_lambdaThetaClimRelax  ADNAME = adffields_lambdaThetaClimRelax
+c$taf COMMON ffields_lambdaThetaClimRelax FTLNAME = g_ffields_lambdaThetaClimRelax
+
+c$taf COMMON ffields_lambdaSaltClimRelax  ADNAME = adffields_lambdaSaltClimRelax
+c$taf COMMON ffields_lambdaSaltClimRelax FTLNAME = g_ffields_lambdaSaltClimRelax
+
+c$taf COMMON DYNVARS_DIFFKR  ADNAME = adDYNVARS_DIFFKR
+c$taf COMMON DYNVARS_DIFFKR FTLNAME = g_DYNVARS_DIFFKR
+
+c$taf COMMON DYNVARS_KAPGM  ADNAME = adDYNVARS_KAPGM
+c$taf COMMON DYNVARS_KAPGM FTLNAME = g_DYNVARS_KAPGM
+
+c$taf COMMON DYNVARS_KAPREDI  ADNAME = adDYNVARS_KAPREDI
+c$taf COMMON DYNVARS_KAPREDI FTLNAME = g_DYNVARS_KAPREDI
+
+c$taf COMMON DYNVARS_BOTTOMDRAG  ADNAME = adDYNVARS_BOTTOMDRAG
+c$taf COMMON DYNVARS_BOTTOMDRAG FTLNAME = g_DYNVARS_BOTTOMDRAG
+
+c$taf COMMON eddypsiffield  ADNAME = adeddypsiffield
+c$taf COMMON eddypsiffield FTLNAME = g_eddypsiffield
+
+c$taf COMMON exf_hsflux_r  ADNAME = adexf_hsflux_r
+c$taf COMMON exf_hsflux_r FTLNAME = g_exf_hsflux_r
+
+c$taf COMMON exf_stress_r  ADNAME = adexf_stress_r
+c$taf COMMON exf_stress_r FTLNAME = g_exf_stress_r
+
+c$taf COMMON exf_wspeed_r  ADNAME = adexf_wspeed_r
+c$taf COMMON exf_wspeed_r FTLNAME = g_exf_wspeed_r
+
+c$taf COMMON exf_atm_temp_r  ADNAME = adexf_atm_temp_r
+c$taf COMMON exf_atm_temp_r FTLNAME = g_exf_atm_temp_r
+
+c$taf COMMON exf_swflux_r  ADNAME = adexf_swflux_r
+c$taf COMMON exf_swflux_r FTLNAME = g_exf_swflux_r
+
+c$taf COMMON exf_atm_wind_r  ADNAME = adexf_atm_wind_r
+c$taf COMMON exf_atm_wind_r FTLNAME = g_exf_atm_wind_r
+
+c$taf COMMON exf_rad_down_r  ADNAME = adexf_rad_down_r
+c$taf COMMON exf_rad_down_r FTLNAME = g_exf_rad_down_r
+
+c$taf COMMON exf_clim_sst_r  ADNAME = adexf_clim_sst_r
+c$taf COMMON exf_clim_sst_r FTLNAME = g_exf_clim_sst_r
+
+c$taf COMMON exf_clim_sss_r  ADNAME = adexf_clim_sss_r
+c$taf COMMON exf_clim_sss_r FTLNAME = g_exf_clim_sss_r
+
+c$taf COMMON seaice_dynvars_1  ADNAME = adseaice_dynvars_1
+c$taf COMMON seaice_dynvars_1 FTLNAME = g_seaice_dynvars_1
+
+c$taf COMMON ggl90_fields  ADNAME = adggl90_fields
+c$taf COMMON ggl90_fields FTLNAME = g_ggl90_fields
+
+c$taf COMMON grid_r  ADNAME = adgrid_r
+c$taf COMMON grid_r FTLNAME = g_grid_r
+
+c$taf COMMON grid_r_c  ADNAME = adgrid_r_c
+c$taf COMMON grid_r_c FTLNAME = g_grid_r_c
+
+c$taf COMMON grid_r_s  ADNAME = adgrid_r_s
+c$taf COMMON grid_r_s FTLNAME = g_grid_r_s
+
+c$taf COMMON grid_r_w  ADNAME = adgrid_r_w
+c$taf COMMON grid_r_w FTLNAME = g_grid_r_w
diff -Naur -x Entries -IHeader MITgcm-c64l/pkg/autodiff/model_ad.flow MITgcm-c64l-v2-patch-final/pkg/autodiff/model_ad.flow
--- MITgcm-c64l/pkg/autodiff/model_ad.flow	2010-12-20 21:04:28.000000000 +0100
+++ MITgcm-c64l-v2-patch-final/pkg/autodiff/model_ad.flow	2013-08-16 00:58:59.000000000 +0200
@@ -1,11 +1,12 @@
-C $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-C $Name:  $
+C $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+C $Name:  $
 
 C----------------------------------------
 C subroutine  the_main_loop
 C----------------------------------------
 CADJ SUBROUTINE the_main_loop  ADNAME = adthe_main_loop
 CADJ SUBROUTINE the_main_loop FTLNAME = g_the_main_loop
+CADJ SUBROUTINE the_main_loop MODNAME = mdthe_main_loop
 
 C----------------------------------------
 C subroutine  calc_oce_mxlayer
diff -Naur -x Entries -IHeader MITgcm-c64l/pkg/autodiff/system.flow MITgcm-c64l-v2-patch-final/pkg/autodiff/system.flow
--- MITgcm-c64l/pkg/autodiff/system.flow	1970-01-01 01:00:00.000000000 +0100
+++ MITgcm-c64l-v2-patch-final/pkg/autodiff/system.flow	2013-08-16 01:01:04.000000000 +0200
@@ -0,0 +1,5 @@
+c$taf subroutine cloc input =
+
+c$taf subroutine csystemtime input =
+
+c$taf subroutine cusertime input =
diff -Naur -x Entries -IHeader MITgcm-c64l/pkg/cost/cost_ad.flow MITgcm-c64l-v2-patch-final/pkg/cost/cost_ad.flow
--- MITgcm-c64l/pkg/cost/cost_ad.flow	1970-01-01 01:00:00.000000000 +0100
+++ MITgcm-c64l-v2-patch-final/pkg/cost/cost_ad.flow	2013-08-16 01:01:13.000000000 +0200
@@ -0,0 +1,12 @@
+C------------------------------------------
+C directives to specify the adjoint names
+C of COMMON blocks used in the original code
+C------------------------------------------
+C$TAF COMMON cost_r FTLNAME = g_cost_r
+C$TAF COMMON cost_r ADNAME = adcost_r
+C$TAF COMMON cost_state_final_r FTLNAME = g_cost_state_final_r
+C$TAF COMMON cost_state_final_r ADNAME = adcost_state_final_r
+C$TAF COMMON cost_vector_r FTLNAME = g_cost_vector_r
+C$TAF COMMON cost_vector_r ADNAME = adcost_vector_r
+C$TAF COMMON cost_dic_cost_ctrl FTLNAME = g_cost_dic_cost_ctrl
+C$TAF COMMON cost_dic_cost_ctrl ADNAME = adcost_dic_cost_ctrl
diff -Naur -x Entries -IHeader MITgcm-c64l/pkg/profiles/profiles.flow MITgcm-c64l-v2-patch-final/pkg/profiles/profiles.flow
--- MITgcm-c64l/pkg/profiles/profiles.flow	2007-10-09 02:07:59.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/pkg/profiles/profiles.flow	2013-08-16 01:01:26.000000000 +0200
@@ -1,5 +1,5 @@
-C $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-C $Name:  $
+C $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+C $Name:  $
 
 C----------------------------------------
 C subroutine profiles_readvector
@@ -7,3 +7,6 @@
 CADJ SUBROUTINE  profiles_readvector   INPUT  =   1,2,3,4,6,7,8
 CADJ SUBROUTINE  profiles_readvector   DEPEND  =  1,2,3,4,6,7
 CADJ SUBROUTINE  profiles_readvector   OUTPUT =   5
+
+CADJ SUBROUTINE  NF_OPEN               INPUT = 
+CADJ SUBROUTINE  NF_CLOSE              INPUT =
diff -Naur -x Entries -IHeader MITgcm-c64l/pkg/seaice/seaice.flow MITgcm-c64l-v2-patch-final/pkg/seaice/seaice.flow
--- MITgcm-c64l/pkg/seaice/seaice.flow	1970-01-01 01:00:00.000000000 +0100
+++ MITgcm-c64l-v2-patch-final/pkg/seaice/seaice.flow	2013-08-16 01:01:41.000000000 +0200
@@ -0,0 +1,9 @@
+C $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+C $Name:  $
+
+C----------------------------------------
+C FUNCTION  SEAICE_DIAG_SUFX
+C----------------------------------------
+CADJ FUNCTION SEAICE_DIAG_SUFX INPUT  = 1,2
+CADJ FUNCTION SEAICE_DIAG_SUFX OUTPUT =
+
diff -Naur -x Entries -IHeader MITgcm-c64l/tools/adjoint_options/adjoint_default MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_default
--- MITgcm-c64l/tools/adjoint_options/adjoint_default	2013-04-04 01:10:54.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_default	2013-08-16 01:02:41.000000000 +0200
@@ -1,10 +1,10 @@
 #!/bin/bash
 #
-#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-#  $Name:  $
+#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+#  $Name:  $
 
-#  This AD option-file contains the default settings for the adjoint and
-#  tangent-linear compilers.  If you need to change these settings,
+#  This AD option-file contains the default settings for the adjoint
+#  and tangent-linear compilers.  If you need to change these settings,
 #  please make a separate (local) copy of this file.
 
 # TAMC=/data43/ralf/tamc/tamc
@@ -15,11 +15,19 @@
 TAMC=tamc
 
 AD_TAMC_FLAGS="-reverse -admark ad -i4 -r4 -l tamc_ad.log $AD_TAMC_FLAGS"
-AD_TAF_FLAGS="-version 2.3.8 -v1 -reverse -admark ad -i4 -r4 -l taf_ad.log $AD_TAF_FLAGS"
 FTL_TAMC_FLAGS="-forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log $FTL_TAMC_FLAGS"
-FTL_TAF_FLAGS="-version 2.3.8 -v1 -forward -ftlmark g_ -i4 -r4 -l taf_ftl.log $FTL_TAF_FLAGS"
 SVD_TAMC_FLAGS="-reverse -forward -pure -i4 -r4 -l tamc_svd.log $SVD_TAMC_FLAGS"
-SVD_TAF_FLAGS="-version 2.3.8 -v1 -reverse -forward -pure -i4 -r4 -l taf_svd.log $SVD_TAF_FLAGS"
+
+AD_TAF_FLAGS="-reverse -server fastopt.net -i4 -r4 -intrinsic system,flush -l taf_ad.log $AD_TAF_FLAGS"
+FTL_TAF_FLAGS="-forward -server fastopt.net -i4 -r4 -intrinsic system,flush -l taf_ftl.log $FTL_TAF_FLAGS"
+SVD_TAF_FLAGS="-reverse -forward -pure -server fastopt.net -i4 -r4 -intrinsic system,flush -l taf_svd.log $SVD_TAF_FLAGS"
+
+#- in case we need to show some MPI code to TAF:
+#if test "x$MPI" != x ; then
+#  AD_TAF_FLAGS="$AD_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#  FLT_TAF_FLAGS="$FLT_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#  SVD_TAF_FLAGS="$SVD_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#fi
 
 DIFF_FLAGS="-toplevel 'the_main_loop'    \
             -input 'xx_theta_dummy       \
diff -Naur -x Entries -IHeader MITgcm-c64l/tools/adjoint_options/adjoint_diva MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_diva
--- MITgcm-c64l/tools/adjoint_options/adjoint_diva	2013-04-04 01:10:54.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_diva	2013-08-16 01:02:41.000000000 +0200
@@ -1,11 +1,11 @@
 #!/bin/bash
 #
-#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-#  $Name:  $
+#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+#  $Name:  $
 #
 
-#  This file contains the default settings for the adjoint and 
-#  tangent-linear compilers.  If you need to change these settings,
+#  This AD option-file contains the default settings for the divided adjoint (DIVA)
+#  and tangent-linear compilers.  If you need to change these settings,
 #  please make a separate (local) copy of this file.
 
 # TAMC=/data43/ralf/tamc/tamc
@@ -16,13 +16,20 @@
 TAF=staf
 TAMC=tamc
 
-
-AD_TAMC_FLAGS='-reverse -admark ad -i4 -r4 -l tamc_ad.log'
-AD_TAF_FLAGS="-version 2.3.8 -v1 -pure -reverse -admark ad -i4 -r4 -l taf_ad.log"
-FTL_TAMC_FLAGS='-forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log'
-FTL_TAF_FLAGS='-version 2.3.8 -v1 -forward -ftlmark g_ -i4 -r4 -l taf_ftl.log'
-SVD_TAMC_FLAGS='-reverse -forward -pure -i4 -r4 -l tamc_svd.log'
-SVD_TAF_FLAGS='-version 2.3.8 -v1 -reverse -forward -pure -i4 -r4 -l taf_svd.log'
+AD_TAMC_FLAGS="-reverse -admark ad -i4 -r4 -l tamc_ad.log $AD_TAMC_FLAGS"
+FTL_TAMC_FLAGS="-forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log $FTL_TAMC_FLAGS"
+SVD_TAMC_FLAGS="-reverse -forward -pure -i4 -r4 -l tamc_svd.log $SVD_TAMC_FLAGS"
+
+AD_TAF_FLAGS="-pure -reverse -i4 -r4 -intrinsic system,flush -l taf_ad.log $AD_TAF_FLAGS"
+FTL_TAF_FLAGS="-forward -i4 -r4 -intrinsic system,flush -l taf_ftl.log $FTL_TAF_FLAGS"
+SVD_TAF_FLAGS="-reverse -forward -pure -i4 -r4 -intrinsic system,flush -l taf_svd.log $SVD_TAF_FLAGS"
+
+#- using DIVA, we need to show some MPI code to TAF:
+if test "x$MPI" != x ; then
+  AD_TAF_FLAGS="$AD_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+  FLT_TAF_FLAGS="$FLT_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+  SVD_TAF_FLAGS="$SVD_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+fi
 
 DIFF_FLAGS="-toplevel 'the_main_loop'    \
             -input 'xx_theta_dummy       \
@@ -51,24 +58,27 @@
                     xx_obcse_dummy       \
                     xx_sst_dummy         \
                     xx_sss_dummy         \
+                    xx_depth_dummy       \
                     xx_diffkr_dummy      \
                     xx_kapgm_dummy       \
-                    xx_edtaux_dummy      \
-                    xx_edtauy_dummy      \
-                    xx_kapredi_dummy       \
+                    xx_kapredi_dummy     \
                     xx_bottomdrag_dummy  \
                     xx_efluxy_dummy      \
                     xx_efluxp_dummy      \
+                    xx_edtaux_dummy      \
+                    xx_edtauy_dummy      \
                     xx_uvel_dummy        \
                     xx_vvel_dummy        \
                     xx_etan_dummy        \
-                    xx_gen2d_dummy        \
-                    xx_gen3d_dummy        \
+                    xx_gen2d_dummy       \
+                    xx_gen3d_dummy       \
                     xx_genarr2d_dummy    \
                     xx_genarr3d_dummy    \
-                    xx_siarea_dummy        \
-                    xx_siheff_dummy        \
-                    xx_sihsnow_dummy        \
+                    xx_gentim2d_dummy    \
+                    xx_siarea_dummy      \
+                    xx_siheff_dummy      \
+                    xx_sihsnow_dummy     \
+                    xx_shifwflx_dummy    \
                     xx_relaxsst_dummy    \
                     xx_relaxsss_dummy    \
                     xx_atemp_mean_dummy  \
@@ -76,7 +86,7 @@
                     xx_precip_mean_dummy \
                     xx_swdown_mean_dummy \
                     xx_uwind_mean_dummy  \
-                    xx_vwind_mean_dummy'     \
+                    xx_vwind_mean_dummy' \
             -output 'fc'"
 
 AD_TAMC_FLAGS="$AD_TAMC_FLAGS $DIFF_FLAGS"
diff -Naur -x Entries -IHeader MITgcm-c64l/tools/adjoint_options/adjoint_f95 MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_f95
--- MITgcm-c64l/tools/adjoint_options/adjoint_f95	2013-04-04 01:10:55.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_f95	2013-08-16 01:02:41.000000000 +0200
@@ -1,11 +1,10 @@
 #!/bin/bash
 #
-#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-#  $Name:  $
-#
+#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+#  $Name:  $
 
-#  This file contains the default settings for the adjoint and 
-#  tangent-linear compilers.  If you need to change these settings,
+#  This file contains the default F95 settings for the adjoint
+#  and tangent-linear compilers.  If you need to change these settings,
 #  please make a separate (local) copy of this file.
 
 # TAMC=/data43/ralf/tamc/tamc
@@ -15,12 +14,20 @@
 TAF=staf
 TAMC=tamc
 
-AD_TAMC_FLAGS='-f95 -reverse -admark ad -i4 -r4 -l tamc_ad.log'
-AD_TAF_FLAGS='-version 2.3.8 -v1 -f95 -reverse -admark ad -i4 -r4 -l taf_ad.log'
-FTL_TAMC_FLAGS='-f95 -forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log'
-FTL_TAF_FLAGS='-version 2.3.8 -v1 -f95 -forward -ftlmark g_ -i4 -r4 -l taf_ftl.log'
-SVD_TAMC_FLAGS='-f95 -reverse -forward -pure -i4 -r4 -l tamc_svd.log'
-SVD_TAF_FLAGS='-version 2.3.8 -v1 -f95 -reverse -forward -pure -i4 -r4 -l taf_svd.log'
+AD_TAMC_FLAGS="-f95 -reverse -admark ad -i4 -r4 -l tamc_ad.log $AD_TAMC_FLAGS"
+FTL_TAMC_FLAGS="-f95 -forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log $FTL_TAMC_FLAGS"
+SVD_TAMC_FLAGS="-f95 -reverse -forward -pure -i4 -r4 -l tamc_svd.log $SVD_TAMC_FLAGS"
+
+AD_TAF_FLAGS="-f95 -reverse -i4 -r4 -intrinsic system,flush -l taf_ad.log $AD_TAF_FLAGS"
+FTL_TAF_FLAGS="-f95 -forward -i4 -r4 -intrinsic system,flush -l taf_ftl.log $FTL_TAF_FLAGS"
+SVD_TAF_FLAGS="-f95 -reverse -forward -pure -i4 -r4 -intrinsic system,flush -l taf_svd.log $SVD_TAF_FLAGS"
+
+#- in case we need to show some MPI code to TAF:
+#if test "x$MPI" != x ; then
+#  AD_TAF_FLAGS="$AD_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#  FLT_TAF_FLAGS="$FLT_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#  SVD_TAF_FLAGS="$SVD_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#fi
 
 DIFF_FLAGS="-toplevel 'the_main_loop'    \
             -input 'xx_theta_dummy       \
@@ -49,9 +56,10 @@
                     xx_obcse_dummy       \
                     xx_sst_dummy         \
                     xx_sss_dummy         \
+                    xx_depth_dummy       \
                     xx_diffkr_dummy      \
                     xx_kapgm_dummy       \
-                    xx_kapredi_dummy       \
+                    xx_kapredi_dummy     \
                     xx_bottomdrag_dummy  \
                     xx_efluxy_dummy      \
                     xx_efluxp_dummy      \
@@ -60,14 +68,15 @@
                     xx_uvel_dummy        \
                     xx_vvel_dummy        \
                     xx_etan_dummy        \
-                    xx_gen2d_dummy        \
-                    xx_gen3d_dummy        \
+                    xx_gen2d_dummy       \
+                    xx_gen3d_dummy       \
                     xx_genarr2d_dummy    \
                     xx_genarr3d_dummy    \
                     xx_gentim2d_dummy    \
-                    xx_siarea_dummy        \
-                    xx_siheff_dummy        \
-                    xx_sihsnow_dummy        \
+                    xx_siarea_dummy      \
+                    xx_siheff_dummy      \
+                    xx_sihsnow_dummy     \
+                    xx_shifwflx_dummy    \
                     xx_relaxsst_dummy    \
                     xx_relaxsss_dummy    \
                     xx_atemp_mean_dummy  \
@@ -75,7 +84,7 @@
                     xx_precip_mean_dummy \
                     xx_swdown_mean_dummy \
                     xx_uwind_mean_dummy  \
-                    xx_vwind_mean_dummy'   \
+                    xx_vwind_mean_dummy' \
             -output 'fc'"
 
 AD_TAMC_FLAGS="$AD_TAMC_FLAGS $DIFF_FLAGS"
diff -Naur -x Entries -IHeader MITgcm-c64l/tools/adjoint_options/adjoint_oad MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_oad
--- MITgcm-c64l/tools/adjoint_options/adjoint_oad	2013-07-07 03:19:22.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_oad	2013-08-16 01:02:41.000000000 +0200
@@ -1,7 +1,7 @@
 #!/bin/bash
 #
 #  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-#  $Name:  $
+#  $Name:  $
 #
 
 #  This file contains settings for OpenAD
diff -Naur -x Entries -IHeader MITgcm-c64l/tools/adjoint_options/adjoint_state_final MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_state_final
--- MITgcm-c64l/tools/adjoint_options/adjoint_state_final	2013-04-04 01:10:55.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_state_final	2013-08-16 01:02:41.000000000 +0200
@@ -1,8 +1,7 @@
 #!/bin/bash
 #
-#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-#  $Name:  $
-#
+#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+#  $Name:  $
 
 #  This file is a template for SVD-type calculations
 #  where a vector-valued cost function (-dependent) is used,
@@ -14,16 +13,24 @@
 # TAF=~fastopt/bin/taf
 # STAF=staf
 
-TAF=taf
+TAF=staf
 TAMC=tamc
 LIBS="${LIBS} -larpack"
 
-AD_TAMC_FLAGS='-reverse -admark ad -i4 -r4 -l tamc_ad.log'
-AD_TAF_FLAGS='-version 2.3.8 -v1 -reverse -admark ad -i4 -r4 -l taf_ad.log'
-FTL_TAMC_FLAGS='-forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log'
-FTL_TAF_FLAGS='-version 2.3.8 -v1 -forward -ftlmark g_ -i4 -r4 -l taf_ftl.log'
-SVD_TAMC_FLAGS='-reverse -forward -pure -i4 -r4 -l tamc_svd.log'
-SVD_TAF_FLAGS='-version 2.3.8 -v1 -reverse -forward -pure -i4 -r4 -l taf_svd.log'
+AD_TAMC_FLAGS="-reverse -admark ad -i4 -r4 -l tamc_ad.log $AD_TAMC_FLAGS"
+FTL_TAMC_FLAGS="-forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log $FTL_TAMC_FLAGS"
+SVD_TAMC_FLAGS="-reverse -forward -pure -i4 -r4 -l tamc_svd.log $SVD_TAMC_FLAGS"
+
+AD_TAF_FLAGS="-reverse -i4 -r4 -intrinsic system,flush -l taf_ad.log $AD_TAF_FLAGS"
+FTL_TAF_FLAGS="-forward -i4 -r4 -intrinsic system,flush -l taf_ftl.log $FTL_TAF_FLAGS"
+SVD_TAF_FLAGS="-reverse -forward -pure -i4 -r4 -intrinsic system,flush -l taf_svd.log $SVD_TAF_FLAGS"
+
+#- in case we need to show some MPI code to TAF:
+#if test "x$MPI" != x ; then
+#  AD_TAF_FLAGS="$AD_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#  FLT_TAF_FLAGS="$FLT_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#  SVD_TAF_FLAGS="$SVD_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#fi
 
 DIFF_FLAGS="-toplevel 'the_main_loop'    \
             -input 'xx_theta_dummy       \
@@ -52,9 +59,10 @@
                     xx_obcse_dummy       \
                     xx_sst_dummy         \
                     xx_sss_dummy         \
+                    xx_depth_dummy       \
                     xx_diffkr_dummy      \
                     xx_kapgm_dummy       \
-                    xx_kapredi_dummy       \
+                    xx_kapredi_dummy     \
                     xx_bottomdrag_dummy  \
                     xx_efluxy_dummy      \
                     xx_efluxp_dummy      \
@@ -63,10 +71,15 @@
                     xx_uvel_dummy        \
                     xx_vvel_dummy        \
                     xx_etan_dummy        \
-                    xx_gen2d_dummy        \
-                    xx_gen3d_dummy        \
+                    xx_gen2d_dummy       \
+                    xx_gen3d_dummy       \
                     xx_genarr2d_dummy    \
                     xx_genarr3d_dummy    \
+                    xx_gentim2d_dummy    \
+                    xx_siarea_dummy      \
+                    xx_siheff_dummy      \
+                    xx_sihsnow_dummy     \
+                    xx_shifwflx_dummy    \
                     xx_relaxsst_dummy    \
                     xx_relaxsss_dummy    \
                     xx_atemp_mean_dummy  \
@@ -74,7 +87,7 @@
                     xx_precip_mean_dummy \
                     xx_swdown_mean_dummy \
                     xx_uwind_mean_dummy  \
-                    xx_vwind_mean_dummy'     \
+                    xx_vwind_mean_dummy' \
             -output 'objf_state_final'"
 
 AD_TAMC_FLAGS="$AD_TAMC_FLAGS $DIFF_FLAGS"
diff -Naur -x Entries -IHeader MITgcm-c64l/tools/adjoint_options/adjoint_tamc_compatibility MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_tamc_compatibility
--- MITgcm-c64l/tools/adjoint_options/adjoint_tamc_compatibility	2012-07-27 20:25:39.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/tools/adjoint_options/adjoint_tamc_compatibility	2013-08-16 01:02:41.000000000 +0200
@@ -1,11 +1,10 @@
 #!/bin/bash
 #
-#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-#  $Name:  $
-#
+#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+#  $Name:  $
 
-#  This file contains the default settings for the adjoint and 
-#  tangent-linear compilers.  If you need to change these settings,
+#  This file contains the TAMC-like settings for the adjoint
+#  and tangent-linear compilers.  If you need to change these settings,
 #  please make a separate (local) copy of this file.
 
 # TAMC=/data43/ralf/tamc/tamc
@@ -15,13 +14,20 @@
 TAF=staf
 TAMC=tamc
 
-AD_TAMC_FLAGS='-reverse -admark ad -i4 -r4 -l tamc_ad.log'
-#AD_TAF_FLAGS='  -reverse -admark ad -i4 -r4 -l taf_ad.log'
-AD_TAF_FLAGS="-v1 -nonew_arg  -reverse -admark ad -i4 -r4 -l taf_ad.log $AD_TAF_FLAGS"
-FTL_TAMC_FLAGS='-forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log'
-FTL_TAF_FLAGS='-v1 -nonew_arg  -forward -ftlmark g_ -i4 -r4 -l taf_ftl.log'
-SVD_TAMC_FLAGS='-reverse -forward -pure -i4 -r4 -l tamc_svd.log'
-SVD_TAF_FLAGS='-v1 -nonew_arg  -reverse -forward -pure -i4 -r4 -l taf_svd.log'
+AD_TAMC_FLAGS="-reverse -admark ad -i4 -r4 -l tamc_ad.log $AD_TAMC_FLAGS"
+FTL_TAMC_FLAGS="-forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log $FTL_TAMC_FLAGS"
+SVD_TAMC_FLAGS="-reverse -forward -pure -i4 -r4 -l tamc_svd.log $SVD_TAMC_FLAGS"
+
+AD_TAF_FLAGS="-v1 -nonew_arg -reverse -i4 -r4 -intrinsic system,flush -l taf_ad.log $AD_TAF_FLAGS"
+FTL_TAF_FLAGS="-v1 -nonew_arg -forward -i4 -r4 -intrinsic system,flush -l taf_ftl.log $FTL_TAF_FLAGS"
+SVD_TAF_FLAGS="-v1 -nonew_arg -reverse -forward -pure -i4 -r4 -intrinsic system,flush -l taf_svd.log $SVD_TAF_FLAGS"
+
+#- in case we need to show some MPI code to TAF:
+#if test "x$MPI" != x ; then
+#  AD_TAF_FLAGS="$AD_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#  FLT_TAF_FLAGS="$FLT_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#  SVD_TAF_FLAGS="$SVD_TAF_FLAGS"' -mpi -include $(LOCAL_MPI_HEADERS) -I./mpi_headers/'
+#fi
 
 DIFF_FLAGS="-toplevel 'the_main_loop'    \
             -input 'xx_theta_dummy       \
@@ -53,7 +59,7 @@
                     xx_depth_dummy       \
                     xx_diffkr_dummy      \
                     xx_kapgm_dummy       \
-                    xx_kapredi_dummy       \
+                    xx_kapredi_dummy     \
                     xx_bottomdrag_dummy  \
                     xx_efluxy_dummy      \
                     xx_efluxp_dummy      \
@@ -62,11 +68,15 @@
                     xx_uvel_dummy        \
                     xx_vvel_dummy        \
                     xx_etan_dummy        \
+                    xx_gen2d_dummy       \
+                    xx_gen3d_dummy       \
                     xx_genarr2d_dummy    \
                     xx_genarr3d_dummy    \
-                    xx_siarea_dummy        \
-                    xx_siheff_dummy        \
-                    xx_sihsnow_dummy        \
+                    xx_gentim2d_dummy    \
+                    xx_siarea_dummy      \
+                    xx_siheff_dummy      \
+                    xx_sihsnow_dummy     \
+                    xx_shifwflx_dummy    \
                     xx_relaxsst_dummy    \
                     xx_relaxsss_dummy    \
                     xx_atemp_mean_dummy  \
@@ -74,7 +84,7 @@
                     xx_precip_mean_dummy \
                     xx_swdown_mean_dummy \
                     xx_uwind_mean_dummy  \
-                    xx_vwind_mean_dummy'   \
+                    xx_vwind_mean_dummy' \
             -output 'fc'"
 
 AD_TAMC_FLAGS="$AD_TAMC_FLAGS $DIFF_FLAGS"
diff -Naur -x Entries -IHeader MITgcm-c64l/tools/genmake2 MITgcm-c64l-v2-patch-final/tools/genmake2
--- MITgcm-c64l/tools/genmake2	2013-07-24 02:33:53.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/tools/genmake2	2013-08-16 01:02:09.000000000 +0200
@@ -1,7 +1,7 @@
 #! /usr/bin/env bash
 #
-# $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-# $Name:  $
+# $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+# $Name:  $
 #
 # Makefile generator for MITgcm UV codes
 #   created  by cnh 03/98
@@ -2703,7 +2703,14 @@
 		      F)
 			echo    " \\"  >> F77srclist.tmp
 			printf " $sf" >> F77srclist.tmp
-			if test "x$OPENAD" != x ; then
+			if test "x$OPENAD" = x ; then
+			    basename=${sf%%.F}
+			    isAD=`egrep ^$basename.f'[ 	]*' adSrcFiles.tmp`
+			    if test -z "$isAD" ; then
+				echo    " \\"  >> nonADF77srclist.tmp
+				printf " $sf" >> nonADF77srclist.tmp
+			    fi
+			else #- OpenAD case:
 			    basename=${sf%%.F}
 			    isAD=`egrep ^$basename.f'[ 	]*' adSrcFiles.tmp`
 			    if test -z "$isAD" ; then
@@ -3136,8 +3143,8 @@
 	ls -l ad_input_code_ad.$FS
 	cat ad_input_code_ad.$FS | sed -f \$(TOOLSDIR)/adjoint_sed > ad_taf_output.$FS
 
-\$(EXE_AD): \$(SPECIAL_FILES) \$(F77_SRC_FILES) \$(C_SRC_FILES) \$(H_SRC_FILES) \$(F90_SRC_FILES) ad_taf_output.o \$(OBJFILES) \$(EMBEDDED_FILES)
-	\$(LINK) -o \${EXE_AD} \$(FFLAGS) \$(FOPTIM) \$(OBJFILES) ad_taf_output.o \$(LIBS)
+\$(EXE_AD): \$(SPECIAL_FILES) \$(H_SRC_FILES) ad_taf_output.o \$(NON_AD_F77_SRC_FILES:.F=.o) \$(F90_SRC_FILES:.F90=.o) \$(C_SRC_FILES:.c=.o) \$(EMBEDDED_FILES)
+	\$(LINK) -o \${EXE_AD} \$(FFLAGS) \$(FOPTIM) ad_taf_output.o \$(NON_AD_F77_SRC_FILES:.F=.o) \$(F90_SRC_FILES:.F90=.o) \$(C_SRC_FILES:.c=.o) \$(LIBS)
 
 ad_tamc_output.$FS: ad_input_code.$FS
 	\$(TAMC) \$(AD_TAMC_FLAGS) \$(TAMC_EXTRA) ad_input_code.$FS
@@ -3176,16 +3183,16 @@
 
 ftl_taf_output.$FS: ftl_input_code.$FS
 	\$(TAF) \$(FTL_TAF_FLAGS) \$(TAF_EXTRA) ftl_input_code.$FS
-	ls -l ftl_input_code_ftl.$FS
-	cat ftl_input_code_ftl.$FS | sed -f \$(TOOLSDIR)/adjoint_sed > ftl_taf_output.$FS
+	ls -l ftl_input_code_tl.$FS
+	cat ftl_input_code_tl.$FS | sed -f \$(TOOLSDIR)/adjoint_sed > ftl_taf_output.$FS
 
 ftltafonly:
 	\$(TAF) \$(FTL_TAF_FLAGS) \$(TAF_EXTRA) ftl_input_code.$FS
-	ls -l ftl_input_code_ftl.$FS
-	cat ftl_input_code_ftl.$FS | sed -f \$(TOOLSDIR)/adjoint_sed > ftl_taf_output.$FS
+	ls -l ftl_input_code_tl.$FS
+	cat ftl_input_code_tl.$FS | sed -f \$(TOOLSDIR)/adjoint_sed > ftl_taf_output.$FS
 
-\$(EXE_FTL): \$(SPECIAL_FILES) \$(F77_SRC_FILES) \$(C_SRC_FILES) \$(H_SRC_FILES) \$(F90_SRC_FILES) ftl_taf_output.o \$(OBJFILES) \$(EMBEDDED_FILES)
-	\$(LINK) -o \${EXE_FTL} \$(FFLAGS) \$(FOPTIM) \$(OBJFILES) ftl_taf_output.o \$(LIBS)
+\$(EXE_FTL): \$(SPECIAL_FILES) \$(H_SRC_FILES) ftl_taf_output.o \$(NON_AD_F77_SRC_FILES:.F=.o) \$(F90_SRC_FILES:.F90=.o) \$(C_SRC_FILES:.c=.o) \$(EMBEDDED_FILES)
+	\$(LINK) -o \${EXE_FTL} \$(FFLAGS) \$(FOPTIM) ftl_taf_output.o \$(NON_AD_F77_SRC_FILES:.F=.o) \$(F90_SRC_FILES:.F90=.o) \$(C_SRC_FILES:.c=.o) \$(LIBS)
 
 ftl_tamc_output.$FS: ftl_input_code.$FS
 	\$(TAMC) \$(FTL_TAMC_FLAGS) \$(TAMC_EXTRA) ftl_input_code.$FS
diff -Naur -x Entries -IHeader MITgcm-c64l/verification/bottom_ctrl_5x5/code_ad/ad_optfile.local MITgcm-c64l-v2-patch-final/verification/bottom_ctrl_5x5/code_ad/ad_optfile.local
--- MITgcm-c64l/verification/bottom_ctrl_5x5/code_ad/ad_optfile.local	2013-04-04 14:29:28.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/verification/bottom_ctrl_5x5/code_ad/ad_optfile.local	2013-08-16 01:02:56.000000000 +0200
@@ -1,7 +1,7 @@
 #!/bin/bash
 #
-#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-#  $Name:  $
+#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+#  $Name:  $
 
 #  This AD option-file contains the settings for the adjoint and
 #  tangent-linear compilers to use for this particular setup.
@@ -16,12 +16,12 @@
 TAF=staf
 TAMC=tamc
 
-AD_TAMC_FLAGS="-reverse -admark ad -i4 -r4 -l tamc_ad.log $AD_TAMC_FLAGS"
-AD_TAF_FLAGS="-version 2.3.8 -v1 -e -reverse -admark ad -i4 -r4 -l taf_ad.log $AD_TAF_FLAGS"
-FTL_TAMC_FLAGS="-forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log $FTL_TAMC_FLAGS"
-FTL_TAF_FLAGS="-version 2.3.8 -v1 -e -forward -ftlmark g_ -i4 -r4 -l taf_ftl.log $FTL_TAF_FLAGS"
+AD_TAMC_FLAGS="-reverse -i4 -r4 -l tamc_ad.log $AD_TAMC_FLAGS"
+AD_TAF_FLAGS="-e -reverse -i4 -r4 -l taf_ad.log $AD_TAF_FLAGS"
+FTL_TAMC_FLAGS="-forward -i4 -r4 -l tamc_ftl.log $FTL_TAMC_FLAGS"
+FTL_TAF_FLAGS="-e -forward -i4 -r4 -l taf_ftl.log $FTL_TAF_FLAGS"
 SVD_TAMC_FLAGS="-reverse -forward -pure -i4 -r4 -l tamc_svd.log $SVD_TAMC_FLAGS"
-SVD_TAF_FLAGS="-version 2.3.8 -v1 -reverse -forward -pure -i4 -r4 -l taf_svd.log $SVD_TAF_FLAGS"
+SVD_TAF_FLAGS="-reverse -forward -pure -i4 -r4 -l taf_svd.log $SVD_TAF_FLAGS"
 
 DIFF_FLAGS="-toplevel 'the_main_loop'    \
             -input 'xx_theta_dummy       \
diff -Naur -x Entries -IHeader MITgcm-c64l/verification/lab_sea/build/genmake_local MITgcm-c64l-v2-patch-final/verification/lab_sea/build/genmake_local
--- MITgcm-c64l/verification/lab_sea/build/genmake_local	2013-07-19 22:44:58.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/verification/lab_sea/build/genmake_local	2013-08-16 01:03:06.000000000 +0200
@@ -4,8 +4,4 @@
 
 #-- This is the local options file for the "new" version of genmake
 
-if test "x$MPI" = x ; then
-  AD_OPTFILE='../../../tools/adjoint_options/adjoint_diva'
-else
-  AD_OPTFILE='../../../tools/adjoint_options/adjoint_diva_mpi'
-fi
+AD_OPTFILE='../../../tools/adjoint_options/adjoint_diva'
diff -Naur -x Entries -IHeader MITgcm-c64l/verification/tutorial_dic_adjoffline/code_ad/ad_optfile.local MITgcm-c64l-v2-patch-final/verification/tutorial_dic_adjoffline/code_ad/ad_optfile.local
--- MITgcm-c64l/verification/tutorial_dic_adjoffline/code_ad/ad_optfile.local	2013-04-04 14:29:28.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/verification/tutorial_dic_adjoffline/code_ad/ad_optfile.local	2013-08-16 01:03:16.000000000 +0200
@@ -1,7 +1,7 @@
 #!/bin/bash
 #
-#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-#  $Name:  $
+#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+#  $Name:  $
 
 #  This AD option-file contains the settings for the adjoint and
 #  tangent-linear compilers to use for this particular setup.
@@ -15,12 +15,12 @@
 TAF=staf
 TAMC=tamc
 
-AD_TAMC_FLAGS="-reverse -admark ad -i4 -r4 -l tamc_ad.log $AD_TAMC_FLAGS"
-AD_TAF_FLAGS="-version 2.3.8 -v1 -reverse -admark ad -i4 -r4 -l taf_ad.log $AD_TAF_FLAGS"
-FTL_TAMC_FLAGS="-forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log $FTL_TAMC_FLAGS"
-FTL_TAF_FLAGS="-version 2.3.8 -v1 -forward -ftlmark g_ -i4 -r4 -l taf_ftl.log $FTL_TAF_FLAGS"
+AD_TAMC_FLAGS="-reverse -i4 -r4 -l tamc_ad.log $AD_TAMC_FLAGS"
+AD_TAF_FLAGS="-reverse -i4 -r4 -l taf_ad.log $AD_TAF_FLAGS"
+FTL_TAMC_FLAGS="-forward -i4 -r4 -l tamc_ftl.log $FTL_TAMC_FLAGS"
+FTL_TAF_FLAGS="-forward -i4 -r4 -l taf_ftl.log $FTL_TAF_FLAGS"
 SVD_TAMC_FLAGS="-reverse -forward -pure -i4 -r4 -l tamc_svd.log $SVD_TAMC_FLAGS"
-SVD_TAF_FLAGS="-version 2.3.8 -v1 -reverse -forward -pure -i4 -r4 -l taf_svd.log $SVD_TAF_FLAGS"
+SVD_TAF_FLAGS="-reverse -forward -pure -i4 -r4 -l taf_svd.log $SVD_TAF_FLAGS"
 
 DIFF_FLAGS="-toplevel 'the_main_loop'    \
             -input 'xx_dic               \
diff -Naur -x Entries -IHeader MITgcm-c64l/verification/tutorial_global_oce_optim/code_ad/ad_optfile.local MITgcm-c64l-v2-patch-final/verification/tutorial_global_oce_optim/code_ad/ad_optfile.local
--- MITgcm-c64l/verification/tutorial_global_oce_optim/code_ad/ad_optfile.local	2013-04-04 14:29:28.000000000 +0200
+++ MITgcm-c64l-v2-patch-final/verification/tutorial_global_oce_optim/code_ad/ad_optfile.local	2013-08-16 01:03:34.000000000 +0200
@@ -1,7 +1,7 @@
 #!/bin/bash
 #
-#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
-#  $Name:  $
+#  $Header: /u/gcmpack/MITgcm/tools/adjoint_options/MITgcm-c64l-TAFv2.patch,v 1.1 2013/08/16 14:53:31 heimbach Exp $
+#  $Name:  $
 
 #  This AD option-file contains the settings for the adjoint and
 #  tangent-linear compilers to use for this particular setup.
@@ -15,12 +15,12 @@
 TAF=staf
 TAMC=tamc
 
-AD_TAMC_FLAGS="-reverse -admark ad -i4 -r4 -l tamc_ad.log $AD_TAMC_FLAGS"
-AD_TAF_FLAGS="-version 2.3.8 -v1 -reverse -admark ad -i4 -r4 -l taf_ad.log $AD_TAF_FLAGS"
-FTL_TAMC_FLAGS="-forward -ftlmark g_ -i4 -r4 -l tamc_ftl.log $FTL_TAMC_FLAGS"
-FTL_TAF_FLAGS="-version 2.3.8 -v1 -forward -ftlmark g_ -i4 -r4 -l taf_ftl.log $FTL_TAF_FLAGS"
+AD_TAMC_FLAGS="-reverse -i4 -r4 -l tamc_ad.log $AD_TAMC_FLAGS"
+AD_TAF_FLAGS="-reverse -i4 -r4 -l taf_ad.log $AD_TAF_FLAGS"
+FTL_TAMC_FLAGS="-forward -i4 -r4 -l tamc_ftl.log $FTL_TAMC_FLAGS"
+FTL_TAF_FLAGS="-forward -i4 -r4 -l taf_ftl.log $FTL_TAF_FLAGS"
 SVD_TAMC_FLAGS="-reverse -forward -pure -i4 -r4 -l tamc_svd.log $SVD_TAMC_FLAGS"
-SVD_TAF_FLAGS="-version 2.3.8 -v1 -reverse -forward -pure -i4 -r4 -l taf_svd.log $SVD_TAF_FLAGS"
+SVD_TAF_FLAGS="-reverse -forward -pure -i4 -r4 -l taf_svd.log $SVD_TAF_FLAGS"
 
 DIFF_FLAGS="-toplevel 'the_main_loop'    \
             -input 'xx_theta_dummy       \
